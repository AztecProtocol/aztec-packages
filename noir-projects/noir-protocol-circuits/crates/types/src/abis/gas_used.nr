use crate::{
    abis::function_selector::FunctionSelector, address::{EthAddress, AztecAddress},
    constants::GAS_USED_LENGTH, hash::pedersen_hash, traits::{Deserialize, Hash, Serialize, Empty},
    abis::side_effect::Ordered, utils::reader::Reader
};

struct GasUsed {
    da_gas: Field,
    l1_gas: Field,
    l2_gas: Field,
}

impl GasUsed {
    fn new(da_gas: Field, l1_gas: Field, l2_gas: Field) -> Self {
        Self { da_gas, l1_gas, l2_gas }
    }

    fn add(self, other: GasUsed) -> Self {
        GasUsed::new(
            self.da_gas + other.da_gas,
            self.l1_gas + other.l1_gas,
            self.l2_gas + other.l2_gas
        )
    }
}

impl Serialize<GAS_USED_LENGTH> for GasUsed {
    fn serialize(self) -> [Field; GAS_USED_LENGTH] {
        [self.da_gas, self.l1_gas, self.l2_gas]
    }
}

impl Deserialize<GAS_USED_LENGTH> for GasUsed {
    fn deserialize(serialized: [Field; GAS_USED_LENGTH]) -> GasUsed {
        GasUsed::new(serialized[0], serialized[1], serialized[2])
    }
}

impl Eq for GasUsed {
    fn eq(self, other : GasUsed) -> bool {
        (self.da_gas == other.da_gas) & (self.l1_gas == other.l1_gas) & (self.l2_gas == other.l2_gas)
    }
}

impl Empty for GasUsed {
    fn empty() -> Self {
        GasUsed::new(0, 0, 0)
    }
}
