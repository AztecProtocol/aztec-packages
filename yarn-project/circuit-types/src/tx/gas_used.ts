import { type Gas } from '@aztec/circuits.js';

export interface GasUsed {
  /**
   * Total gas used across both private and public executions.
   * Note that this does not determine the transaction fee. The fee is calculated using `teardownGasLimits` from
   * `GasSettings`, rather than actual teardown gas.
   */
  totalGas: Gas;
  /**
   * The actual gas used in the teardown phase.
   */
  teardownGas: Gas;
}
