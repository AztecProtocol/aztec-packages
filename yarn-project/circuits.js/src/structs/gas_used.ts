import { BufferReader } from '@aztec/foundation/serialize';

import { inspect } from 'util';

export class GasUsed {
  public static readonly PACKED_SIZE_IN_BYTES = 8n;
  public static readonly MAX_VALUE = 2n ** (GasUsed.PACKED_SIZE_IN_BYTES * 8n) - 1n;
  private static readonly PREIMAGE_SIZE_IN_BYTES = 32n;

  public readonly value: bigint;

  constructor(gas: bigint) {
    if (gas < 0) {
      throw new Error('Gas used cannot be negative');
    } else if (gas > GasUsed.MAX_VALUE) {
      throw new Error(`Gas used is too large: [${gas}] does not fit in a ${GasUsed.PACKED_SIZE_IN_BYTES} byte field`);
    }

    this.value = BigInt(gas);
  }

  public toHashPreimage(): Buffer {
    const padding = Buffer.alloc(Number(GasUsed.PREIMAGE_SIZE_IN_BYTES - GasUsed.PACKED_SIZE_IN_BYTES));
    return Buffer.concat([padding, this.toBuffer()]);
  }

  public toBuffer(): Buffer {
    const b = Buffer.alloc(Number(GasUsed.PACKED_SIZE_IN_BYTES));
    b.writeBigUInt64BE(this.value);
    return b;
  }

  public equals(other: GasUsed): boolean {
    return this.value === other.value;
  }

  static empty(): GasUsed {
    return new GasUsed(0n);
  }

  static fromBuffer(buffer: Buffer | BufferReader): GasUsed {
    const reader = BufferReader.asReader(buffer);
    const gas = reader.readBytes(Number(GasUsed.PACKED_SIZE_IN_BYTES)).readBigUInt64BE();
    return new GasUsed(gas);
  }

  /**
   *
   * @returns A barely random instance of GasUsed. Not suitable for cryptographic use.
   */
  static random(): GasUsed {
    let g = GasUsed.MAX_VALUE + 1n;
    while (g > GasUsed.MAX_VALUE) {
      g = BigInt(Math.floor(Math.random() * Number(GasUsed.MAX_VALUE)));
    }
    return new GasUsed(g);
  }

  [inspect.custom]() {
    return `GasUsed<${this.value}>`;
  }
}
