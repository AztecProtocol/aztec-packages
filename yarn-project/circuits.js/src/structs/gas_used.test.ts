import { GasUsed } from './gas_used.js';

describe('gas_used', () => {
  it.each([0n, 1n, 2n, 3n, 42n, 2n ** 32n, 2n ** 64n - 1n])('accepts valid values', value => {
    const gasUsed = new GasUsed(value);
    expect(gasUsed.value).toEqual(BigInt(value));

    const b = gasUsed.toBuffer();
    expect(b).toMatchSnapshot();
    expect(GasUsed.fromBuffer(b)).toEqual(gasUsed);

    expect(gasUsed.toHashPreimage()).toMatchSnapshot();
  });

  it.each([-1n, -2n, -42n])('rejects negative values', value => {
    expect(() => new GasUsed(value)).toThrow();
  });

  it('rejects too large values', () => {
    expect(() => new GasUsed(2n ** 64n)).toThrow();
  });

  it('serde retains value', () => {
    const a = GasUsed.fromBuffer(new GasUsed(42n).toBuffer());
    const b = GasUsed.fromBuffer(new GasUsed(43n).toBuffer());

    expect(a.value).toEqual(42n);
    expect(b.value).toEqual(43n);

    expect(a.equals(a)).toBeTruthy();
    expect(a.equals(b)).toBeFalsy();
  });

  it('random', () => {
    const g = GasUsed.random();
    expect(g.value).toBeLessThanOrEqual(GasUsed.MAX_VALUE);
  });
});
