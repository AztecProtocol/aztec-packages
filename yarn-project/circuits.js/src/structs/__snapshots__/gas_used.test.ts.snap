// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gas_used accepts valid values 1`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`gas_used accepts valid values 2`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`gas_used accepts valid values 3`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
  ],
  "type": "Buffer",
}
`;

exports[`gas_used accepts valid values 4`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
  ],
  "type": "Buffer",
}
`;

exports[`gas_used accepts valid values 5`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    2,
  ],
  "type": "Buffer",
}
`;

exports[`gas_used accepts valid values 6`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    2,
  ],
  "type": "Buffer",
}
`;

exports[`gas_used accepts valid values 7`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    3,
  ],
  "type": "Buffer",
}
`;

exports[`gas_used accepts valid values 8`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    3,
  ],
  "type": "Buffer",
}
`;

exports[`gas_used accepts valid values 9`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    42,
  ],
  "type": "Buffer",
}
`;

exports[`gas_used accepts valid values 10`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    42,
  ],
  "type": "Buffer",
}
`;

exports[`gas_used accepts valid values 11`] = `
{
  "data": [
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`gas_used accepts valid values 12`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
  ],
  "type": "Buffer",
}
`;

exports[`gas_used accepts valid values 13`] = `
{
  "data": [
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
  ],
  "type": "Buffer",
}
`;

exports[`gas_used accepts valid values 14`] = `
{
  "data": [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
    255,
  ],
  "type": "Buffer",
}
`;
